docs/DafnyRef/Types.md

```diff
@@ -1172,21 +1203,248 @@ can be used abstractly to represent an arbitrary parameterized monad.
+base type.  The newtype is distinct from and incompatible with other
+types; in particular, it is not assignable to its base type
+without an explicit conversion.  An important difference between the
+operations on a newtype and the operations on its base type is that
+the newtype operations are defined only if the result satisfies the
+predicate `Q`, and likewise for the literals of the
+newtype.
+
+For example, suppose `lo` and `hi` are integer-based numerics that
+satisfy `0 <= lo <= hi` and consider the following code fragment:
+```dafny
+var mid := (lo + hi) / 2;
+```
+If `lo` and `hi` have type `int`, then the code fragment is legal; in
+particular, it never overflows, since `int` has no upper bound.  In
+contrast, if `lo` and `hi` are variables of a newtype `int32` declared
+as follows:
+```dafny
+newtype int32 = x | -0x80000000 <= x < 0x80000000
+```
+then the code fragment is erroneous, since the result of the addition
+may fail to satisfy the predicate in the definition of `int32`.  The
+code fragment can be rewritten as
+```dafny
+var mid := lo + (hi - lo) / 2;
+```
+in which case it is legal for both `int` and `int32`.
+
+Since a newtype is incompatible with its base type and since all
+results of the newtype's operations are members of the newtype, a
+compiler for Dafny is free to specialize the run-time representation
+of the newtype.  For example, by scrutinizing the definition of
+`int32` above, a compiler may decide to store `int32` values using
+signed 32-bit integers in the target hardware.
+
+The incompatibility of a newtype and its basetype is intentional,
+as newtypes are meant to be used as distinct types from the basetype.
+If numeric types are desired that mix more readily with the basetype,
+the subset types described in [Section 11.3](#sec-subset-types)
+ may be more appropriate.
+
+Note that the bound variable `x` in `Q` has type `M`, not `N`.
+Consequently, it may not be possible to state `Q` about the `N`
+value.  For example, consider the following type of 8-bit 2's
+complement integers:
+```dafny
+newtype int8 = x: int | -128 <= x < 128
+```
+and consider a variable `c` of type `int8`.  The expression
+```dafny
+-128 <= c < 128
+```
+is not well-defined, because the comparisons require each operand to
+have type `int8`, which means the literal `128` is checked to be of
+type `int8`, which it is not.  A proper way to write this expression
+would be to use a conversion operation, described in [Section 12.1](#sec-conversion), on `c` to
+convert it to the base type:
+```dafny
+-128 <= c as int < 128
+```
+
+If possible, Dafny compilers will represent values of the newtype using
+a native data type for the sake of efficiency. This action can
+be inhibited or a specific native data type selected by
+using the `{:nativeType}` attribute, as explained in
+[Section 22.1.12](#sec-nativetype).
+
+There is a current restriction that the value `0` must be part of every
+numeric newtype.
```

Comment:
This is no longer true. The `witness` clause is used to override this.

Reply:
Removed